Trazendo a image gerada no codenarc-action

jemersonmaia opened the pull request:
Transferindo criação da imagem que estava no projeto `codenarc-action` para o projeto `codenarc-docker`, pois isso vai permitir construir a imagem antes de ser utilizada no `codenarc-action`, melhorando a performance em sua utilização.

Com o objetivo de facilitar os testes, o workflow ci.yml fará publicação de uma imagem ao realizar pull request (com o número da PR) e ao criar uma release manual (podendo ser versão beta, snapshot, rc, etc).

Já o workflow release.yml, que utiliza [bumpr](https://github.com/haya14busa/action-bumpr) fará publicação da imagem latest e última versão gerado ao confirmar (merge) a PR.

O action.yml do projeto `codenarc-action` poderá passar a ter a configuração:

```
runs:
  using: 'docker'
  image: 'docker://docker.io/asaasdev/codenarc:latest' # ou a última versão fixa
```

jemersonmaia committed:
Trazendo a image gerada no codenarc-action

jemersonmaia committed:
Ajustando recomandação hadolint

luiszimmermann commented on .github/dependabot.yml:
```diff
@@ -5,10 +5,6 @@
 
 version: 2
 updates:
-  - package-ecosystem: "github-actions" # See documentation for possible values
```
Tem algum motivo pra remoção do `dependabot`?

jemersonmaia replied:
Aqui eu removi essa parte:

```
- package-ecosystem: "github-actions" # See documentation for possible values
    directory: "/" # Location of package manifests
    schedule:
      interval: "daily"
``` 

Ela não é usada só quando o projeto é do tipo `github-actions`?

luiszimmermann replied:
Entendi, mas não, ela é usada para verificar as versões das actions usadas no repositório.

jemersonmaia replied:
Legal! Corrigido.

luiszimmermann commented on .github/workflows/ci.yml:
```diff
@@ -1,34 +1,39 @@
+      - name: Build and push
         id: docker_build
-        uses: docker/build-push-action@v2.7.0
+        uses: docker/build-push-action@v2.6.1
```
Voltou a versão para a `2.6.1`

jemersonmaia replied:
Corrigido

luiszimmermann commented on .github/workflows/ci.yml:
```diff
@@ -1,34 +1,39 @@
     runs-on: ubuntu-latest
     steps:
-      -
-        name: Set up QEMU
```
Essa parte do `QEMU` e do `buildx` é importante pois ele constrói uma imagem agnóstica que pode rodar em runtimes que não sejam o docker com total compatibilidade. É a maneira mais nova e correta de construir imagens.

jemersonmaia replied:
Entendi. Voltei essa configuração.

luiszimmermann commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,87 @@
+        with:
+          push: true
+          tags: |
+            jfnandopr/codenarc:${{ steps.get-latest-tag.outputs.tag }}
```
Precisa ajustar para o repositório do `asaasdev`

jemersonmaia replied:
Corrigido

luiszimmermann commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,87 @@
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}      
+
+      - name: Build and push
```
Aqui é necessário colocar aquele esquema do `QEMU` e `buildx`

jemersonmaia replied:
Adicionado.

jemersonmaia committed:
Corrigida versão docker/build-push-action
Voltadas etapas Set up QEMU e  Set up Docker Buildx
Corrigido repositório Docker Hub para asaasdev/codenarc

jemersonmaia committed:
Corrigido dependabot

luiszimmermann reviewed (approved):
LGTM,
Mas eu tenho uma dúvida, a action do ci terá acesso aos secrets necessários em novos PRs para realizar o push no docker hub? Se sim, seria possível um PR alterar o ci e sobreescrever a imagem `latest`?

jemersonmaia commented:
> LGTM, Mas eu tenho uma dúvida, a action do ci terá acesso aos secrets necessários em novos PRs para realizar o push no docker hub? Se sim, seria possível um PR alterar o ci e sobreescrever a imagem `latest`?

O ci tem acesso ao secrets e dependendo da configuração alterada no ci.yml poderia sobrescrever sim. A configuração atual é sem `flavor`, onde [por padrão o latest é `auto`](https://github.com/docker/metadata-action#flavor-input):

```
- name: Docker meta
  id: meta
  uses: docker/metadata-action@v3
  with:          
    images: |
      asaasdev/codenarc
    tags: |
      type=ref,event=pr
      type=semver,pattern={{version}}
```

Caso seja alterado para o exemplo abaixo, sempre mandaria a latest, mesmo sem uma PR:

```
- name: Docker meta
  id: meta
  uses: docker/metadata-action@v3
  with:          
    images: |
      asaasdev/codenarc
    tags: |
      type=ref,event=pr
      type=semver,pattern={{version}}
    flavor: |
      latest=true
```

jemersonmaia commented:
> > LGTM, Mas eu tenho uma dúvida, a action do ci terá acesso aos secrets necessários em novos PRs para realizar o push no docker hub? Se sim, seria possível um PR alterar o ci e sobreescrever a imagem `latest`?
> 
> O ci tem acesso ao secrets e dependendo da configuração alterada no ci.yml poderia sobrescrever sim. A configuração atual é sem `flavor`, onde [por padrão o latest é `auto`](https://github.com/docker/metadata-action#flavor-input):
> 
> ```
> - name: Docker meta
>   id: meta
>   uses: docker/metadata-action@v3
>   with:          
>     images: |
>       asaasdev/codenarc
>     tags: |
>       type=ref,event=pr
>       type=semver,pattern={{version}}
> ```
> 
> Caso seja alterado para o exemplo abaixo, sempre mandaria a latest, mesmo sem uma PR:
> 
> ```
> - name: Docker meta
>   id: meta
>   uses: docker/metadata-action@v3
>   with:          
>     images: |
>       asaasdev/codenarc
>     tags: |
>       type=ref,event=pr
>       type=semver,pattern={{version}}
>     flavor: |
>       latest=true
> ```

@luiszimmermann, vou precisar alterar a configuração 

```
tags: |
    type=ref,event=pr
    type=semver,pattern={{version}}
```

para 

```
tags: |
    type=ref,event=pr
    type=semver,pattern={{raw}}
```

Isso porque são dois actions diferentes que obtêm a versão da release (tags):

- [docker/metadata-action@v3 do ci.yml](https://github.com/docker/metadata-action)
- [actions-ecosystem/action-get-latest-tag@v1 no release.yml](https://github.com/marketplace/actions/actions-ecosystem-action-get-latest-tag)

Não foi possível usar o mesmo do ci no release pois o release usa bumpr pra gerar nova release de forma automática, o que não é pega no docker/metadata-action@v3. Porém, ambas tratam versão de formas diferentes. O action-get-latest-tag não remove o **v** da tag e não tem uma forma de configurar isso. Já no metadata-action é possível configurar com o `raw` onde ele passará a usar o **v** em suas versões também.

Só destacando que o metadata-action usará a tag pra gerar uma nova imagem somente se for criado uma release manualmente. Caso contrário será utilizada somente no PR. Quem deve liberar a versão final e a latest é o release.yml após o bumpr criar a release de forma automática.

jemersonmaia committed:
Ajustado pattern do docker/metadata-action@v3 para obter versão igual a obtida pelo  actions-ecosystem/action-get-latest-tag@v1
